src/experiments/data/registry.js

```diff
@@ -0,0 +1,126 @@
+import TextInput from '../components/field-text-input';
+import ImageInput from '../components/field-image-input';
+
+const { doAction } = wp.hooks;
+
+/**
+ * Tests registry class.
+ */
+export default class TestsRegistry {
+	tests = {};
+
+	/**
+	 * Initiatlize the class, transforming passed options to usable test objects.
+	 *
+	 * @param {object} tests Initial tests object to see the registry.
+	 */
+	constructor( tests ) {
```

Comment:
You're going to hate me for this but I'm not a fan of this approach to creating the registry.

- All post A/B tests will be registered in the sidebar UI - we explicitly don't want this for the A/B testing block so it should be opt in
- Having to create / expose a custom React component on the `window` object is a bit clunky, why not allow it to be passed when registering a test on the client side
- Most registry implementations use a `register` type method rather than updating existing items, this doesn't feel idiomatically similar to other existing APIs in WordPress

I preferred the opt-in approach with the `registerExperiment` function. It matched the way blocks are registered on the server and client side and meant it would be really straightforward to provide a custom component etc... I had in mind something along these lines:

```js
Altis.Analytics.Experiments.registerExperiment = ( id, options ) => {
   if ( ! Altis.Analytics.Experiments.PostABTests[ id ] ) {
     throw new Error( 'Test must be registered server side ... etc...' );
   }

   // Validate options...

   Altis.Analytics.Experiments.PostABTests[ id ] = {
     ...Altis.Analytics.Experiments.PostABTests[ id ],
     ...options,
     showUI: true,
   };
}
```

then to use it:

```js
const { registerExperiment } = Altis.Analytics.Experiments;

registerExperiment( 'titles', {
   component: TextInput,
   dispatcher( ( dispatch, values ) => { ... } ),
   ...
} );
```

Then check which AB tests have a component or the example flag like `showUI` or maybe `showInSidebar` to build the plugin sidebar.

It's a rough example but I think that kind of thing would to be a lot easier to work with than passing `input_type` from the backend or exposing react components on `window`.

Could still use a hook to pass a more contained registry object through if you prefer, doesn't have to be on the Altis window object.

Reply:
> I still prefer to use input_type to make the front-end registration optional. Maybe even go for field registration, which would make it even more extensible/reusable.

I'm not really convinced about setting this server side still, I don't want to have to maintain a bunch of built in field types when it's highly likely a custom dispatcher & selector would be needed anyway. We'd potentially be getting into passing all sorts of other options for a dropdown or field type that needs extra config, It'll get complicated quickly and it's more documentation to write and maintain.

I think the client side aspects will be easier to deal with purely on the client side and let implementors figure out what they need.

> I tried to get out of using globals for the test definitions object, but it was either a global ( eg: a registry of some sort / a global var ), or a filter, can also go for a context, something I hadn't thought of until now, and it makes quite more sense!

I'd probably leave a refactor like that for a later time tbh, there are few different ways/places that global data is used and was just simpler to understand and work with initially, especially once I had to start splitting out files so they could be seaprately enqueued. Globals aren't strictly a bad thing, they're easy to understand and can be useful - thinking of `wp.apiFetch`, `wp.i18n`, `wp.components.Thing` etc...

> I had some hard time following the terminology, I had started using experiment, then test, then abtest, then postabtest, trying to differentiate between the current test object that has the results and is being used everywhere, to avoid conflicts / confusion. We may need some refactoring there to make it less confusing. It'd be great if we can align now on all the terms and what they mean.

Yeah again, there was a fair bit of code churn and figuring things out as I went when I originally wrote it. "Post A/B Test" was named because a test was specifically associated with a post in the db. I originally wrote it so you could register different data storage and retrieval functions with a test but it was over engineered.

We can fix this in a later update I think for a v5. If it helps the following things are true:

- Experiments is a blanket term that gives us flexibility for different types of tests we add in future
- An A/B test is one type of experiment, a Multivariate test is another
- A test could be associated with a post, or potentially with the site or other entity (will see if the need for that actually comes up)

Hopefully that clears it up a bit.

In terms of how to proceed:

> - Use a `show_ui` prop to decide which experiments to add to the sidebar, during registration of the test/experiment.

✅  can use this to filter them out server side.

> - Use a context to pass down the JSON object, instead of the current kinda-registry approach. And to register custom components as well.

I don't fully get what advantage the context has, does it just make it more testable in isolation? I guess you'd do something like `useContext( 'experiments' )`? Can we add the context provider high enough or would we have to create it twice, like for the plugin icon and the plugin sidebar?

Again I wouldn't really go with the `input_type` option on the backend myself, I think there's always going to be some client side code to handle the dispatcher and selector and that's going to be specific per test rather than per field. Registering field types is getting a bit abstract in this instance as we're not making a form system.

Whichever way you go, as long as defining the field component for a test is easy and intuitive that's the main thing, a filter or function call to modify the test object would be simplest IMO. I'll leave it in your hands.

> - Use a JS event ( or a WP action/filter ) during panel registration to allow 3rd party code to adjust/override values.

✅ 

> - Use a PHP WP action at the end of register_post_ab_test() to allow 3rd party code to filter the registered test.

✅ 

> - Get rid of the post types support filter and use WP post type support functions all the way.

✅

Reply:
The data inheritance here is perfect for React context, specially as with the `withTestData` HOC, it all becomes quite seamless.

46482b5 does the following:

- Use React Context to manage data inheritance, within `withTestData`, no prop passing anymore, each panel gets its own `abTest` prop if wrapped with `withTestData`, with available exported `context` object for any component ( eg the input ) usable with `useContext()` if needed.
- Use `show_ui` prop to skip tests with no sidebar UI.
- Remove `input_type` in favor of front-end explicit registration.
- Added a filter to manage registered test options before any handling, and an action to fire anything after the test is registered.
- Moved the REST API field registration to a separate hook on `rest_api_init` so new filters/actions (from the above) can action changes (like post type support) and have that accounted for there.
- Moved to use `get_post_type_by_support` all over.